Refactoring.

Stan-Kudri opened the pull request:
-Removed the field "_tail";
-Removed actions that the method should not perform "Intersect"; -Property "NextPoint" is called in methods "EatFood" and "Move"; -Add private method "ClampInvert";
-Add methods for the "_body" "Draw" and "Clear";
-Removed actions that the method should not perform "BuildBody";

Stan-Kudri committed:
Changed class Snake:
-Removed the field "_tail";
-Removed actions that the method should not perform "Intersect";
-Property "NextPoint" is called in methods "EatFood" and "Move";
-Add private method "ClampInvert";
-Add methods for the "_body" "Draw" and "Clear";
-Removed actions that the method should not perform "BuildBody";

Stan-Kudri committed:
Changed class "Field":
-Replaced the actions of methods "HorizontalBorder" and "VerticalBorder" with the method "Border";
-Add method "Draw".

Stan-Kudri committed:
Changed:

-Enum "Directions": Rename from "Other" to "Unknown" and changed position 4 -> 0;
-Class "Directions": Rename from "Turn" to "Value";
-Program: Renamed the parameter from "directory" to "direction".

Stan-Kudri committed:
Rename:
-Class "Food": SetPosition -> GeneratePosition;
-Main: Method from "DisplaySettings" -> "WindowSetting".

Stan-Kudri committed:
Chenged class "Score":
-Add const "StartWidthDisplay";
-Input parameter "height";
-Changed constructor;
-Add method "Add" for property "Points".

Stan-Kudri committed:
Rename and Changed:

-Rename class from "Field" to "Border";
-Class "Score": Rename method from "Add" to "Interease" and Removed the constant "TitleScoreLine";
-Class "Snake": Deleted the variable "nextHeadPosition" in method "EatFood".

Stan-Kudri committed:
Big Changed:

-Add to folder "Components" to folder "ItemGameMap" and moved it there class "Border", "Food" and "Snake";
-Changed classes "Border", "Food" and "Snake";
-Added a class "GameMap" that stores class instances "Border", "Food" and "Snake";
-Changed main class.

blowin commented on Snake/Components/GameMap.cs:
```diff
@@ -0,0 +1,48 @@
+
+        public void ChangeSnakeDirection(Direction direction) => _snake.Direction = direction.Value;
+
+        public void Movie(out int scoreFood) {
```
Что с названием, почему out параметр?

blowin commented on Snake/Components/ItemGameMap/Snake.cs:
```diff
@@ -1,57 +1,66 @@
-                    foreach (var point in _body) {
-                        point.Clear();
-                    }
+        private Point NextPoint {
```
Сделай метод а не свойство

blowin commented on Snake/Components/ItemGameMap/Border.cs:
```diff
@@ -0,0 +1,38 @@
+
+        public void Draw() => _borders.ForEach(x => x.Draw(SymbolField));
+
+        private void Create() {
```
пусть возвращает List<Point>

blowin commented on Snake/Program.cs:
```diff
@@ -7,68 +7,64 @@
+        gameMap.Movie(out var scoreFood);
+        gameMap.Draw();
+
+        score.Increase(scoreFood);
```
привяжи лучше scoreFood + score.Increase через событие

Stan-Kudri committed:
Changes by comments:

-Class "Border":
  1)Rename the field from "_borders" to "_border";
  2)The "Create" method return a list of border elements "List<Point>";
  3)The "_border" field in the constructor takes the value of the "Create" method.

-Class "Snake":
  1)Changed the "Length" property into a private "_length" field;
  2)Changed from property to method "NextPoint".

-Class "GameMap":
  1)Add event for eat food Action<Food>? OnEatScore ();
  2)Add method "HeightBoard" -> Board.Height;

-Class "Score":
  1)The constructor takes a "GameMap" as input and subscribes to the eaten event;
  2)Chenged method Increase.

-Main Chenged:I Changed the main form according to new changes in other classes.

blowin commented on Snake/Components/GameMap.cs:
```diff
@@ -1,48 +1,59 @@
+            if (_snake.EatFood(_food.Position))
+            {
+                OnEatScore?.Invoke(_food);
+                while (_snake.IntersectBody(_food.Position))
```
надо оптимальнее делать, если 3 раза не нашло случайно точку, иди по карте и в первое свободное место вставлять, сделай метод, который возвращает новую точку для еды. Так как у тебя событие есть, можно подумать о том, чтобы вынести логику генерации еды в отдельное место

blowin commented on Snake/Components/ItemGameMap/Snake.cs:
```diff
@@ -73,9 +52,12 @@ public class Snake {
-                if (_head.Equals(_body[i])) {
+        public bool Intersect()
+        {
+            for (var i = 1; i < _length - 1; i++)
```
а зачем начинать с последнего элемента?

blowin commented on Snake/Components/Score.cs:
```diff
@@ -1,27 +1,29 @@
+        public Score(GameMap map)
+        {
+            _startHeightDisplay = map.HeightBoard() + 2;
+            map.OnEatScore += Increase;
```
это что за новости? Пусть кто-то кто создаёт Score подпишет его на событие в GameMap, не передавай сюда карту

Stan-Kudri committed:
Changed the class instance signature to an event.

Stan-Kudri committed:
Moved the logic of the "Game" and "GameLoop" methods to the "Game" class with 2 methods.

Stan-Kudri committed:
Changed:

-Main: Add two empty lines.

-Changed the class from "Direction" to "UserInput":
  1)Renamed the class;
  2)Add event Action<UserInput>? OnChangedDirection;
  3)Added the "UseKey" method with a piece of the logic of the "Run" method of the "Game" class;
  4)Made the "ChangeDirection" method private.

-Class "GameMap":Remove method "HeightBoard".

-Class "Game": Removed part of the logic of the "Run" method, which was transferred to the "UseKey" method of the "UserInput" class.

Stan-Kudri committed:
Add a search for an empty cell for food, if after three searches for a random cell return false.

Stan-Kudri committed:
Changed and  Fixed Bug (Snake):
Fixed a bug in the method and sorted through the values of the sheet from the end.

blowin commented on Snake/Components/GameMap.cs:
```diff
@@ -21,19 +23,17 @@ public GameMap(int width, int height, int snakeLength = 10)
+                if (!RandomSearchCellForFood())
                 {
-                    _food = new Food(_border.GenerateFoodPosition());
+                    SearchCellForFood();
```
почему метод ничего не возвращает?

blowin commented on Snake/Components/Score.cs:
```diff
@@ -8,10 +8,9 @@ public class Score
         {
-            _startHeightDisplay = map.HeightBoard() + 2;
-            map.OnEatScore += Increase;
+            _startHeightDisplay = height + 2;
```
передавай стартовые очки

blowin commented on Snake/Game.cs:
```diff
@@ -0,0 +1,51 @@
+
+        public void Run()
+        {
+            while (!GameOver)
```
почему бы не сделать _gameMap.GameOver

blowin commented on Snake/Game.cs:
```diff
@@ -0,0 +1,51 @@
+    {
+        private bool GameOver = false;
+        private GameMap _gameMap;
+        private UserInput _directory;
```
что такое directory?

blowin commented on Snake/Program.cs:
```diff
@@ -1,30 +1,19 @@
-    var gameMap = new GameMap(width, height);
-    var score = new Score(gameMap);
+var game = new Game(width, height);
+game.DrawMap();
```
почему Game не может это в Run сделать?

blowin commented on Snake/UserInput.cs:
```diff
@@ -0,0 +1,51 @@
+
+        public Directions CurrentDirection { get; private set; }
+
+        public void UseKey()
```
может Update или как-то ещё, но название мне не нравится

blowin commented on Snake/Components/GameMap.cs:
```diff
@@ -55,5 +55,38 @@ public void Clear()
             _food.Clear();
         }
+
+        private bool RandomSearchCellForFood()
```
почему bool а не Point? ??

Stan-Kudri committed:
Changed:

-Class "GameMap":
  1)Remove method "DrawBoarder()";
  2)Rename method from "RandomSearchCellForFood" to "RandomCellForFood";
  3)Changed the output parameters of the "RandomCellForFood" and "SearchCellForFood" methods from "bool" to "Food?";
  4)Changed method "Move".

-Class "Score":Add an input parameter "int points" to the constructor.

-Class "Game":
  1)Rename field from "_deretory" to "_direction";
  2)Changed the logic in the "Run" method.
  3)Remove method DrawMap();

-Class "UserInput": Rename method from "UseKey" to "Update".

blowin commented on Snake/Components/GameMap.cs:
```diff
@@ -30,22 +30,17 @@ public void Move()
-                {
-                    SearchCellForFood();
-                }
+                _food = RandomCellForFood() ?? SearchCellForFood() ?? throw new Exception("There is no empty cell for food.");
```
выглядит жестко, но лучше)

blowin commented on Snake/Components/Score.cs:
```diff
@@ -8,9 +8,10 @@ public class Score
-        public Score(int height)
+        public Score(int height, int points = 0)
         {
             _startHeightDisplay = height + 2;
```
назови константу 2, что это такое?

blowin commented on Snake/Game.cs:
```diff
@@ -4,48 +4,35 @@ namespace GameSnake
-        private bool GameOver = false;
         private GameMap _gameMap;
-        private UserInput _directory;
+        private UserInput _direction;
```
наверное это userInput

Stan-Kudri committed:
Changed:
-Class "Game": Rename field from "_direction" to '_userInput".
Class "Score": Add constant "OffsetPositionHeight" and applied it constructor.
